src/logics/LALogic.cc

```diff
@@ -122,25 +122,25 @@ lbool LALogic::arithmeticElimination(const vec<PTRef> & top_level_arith, MapWith
             // Already has substitution for this var, let the main substitution code deal with this situation
             continue;
         } else {
-            substitutions.insert(var, PtAsgn(sub.second, l_True));
+            substitutions.insert(var, sub.second);
         }
     }
     // To simplify this method, we do not try to detect a conflict here, so result is always l_Undef
-    return l_Undef;
+    return {l_Undef, std::move(substitutions)};
 }
 
-lbool LALogic::retrieveSubstitutions(const vec<PtAsgn>& facts, MapWithKeys<PTRef,PtAsgn,PTRefHash>& substs)
+opensmt::pair<lbool,Logic::SubstMap> LALogic::retrieveSubstitutions(const vec<PtAsgn>& facts)
 {
-    lbool res = Logic::retrieveSubstitutions(facts, substs);
-    if (res != l_Undef) return res;
+    auto resAndSubsts = Logic::retrieveSubstitutions(facts);
+    if (resAndSubsts.first != l_Undef) return resAndSubsts;
     vec<PTRef> top_level_arith;
     for (int i = 0; i < facts.size(); i++) {
         PTRef tr = facts[i].tr;
         lbool sgn = facts[i].sgn;
         if (isNumEq(tr) && sgn == l_True)
             top_level_arith.push(tr);
     }
-    return arithmeticElimination(top_level_arith, substs);
+    return arithmeticElimination(top_level_arith, std::move(resAndSubsts.second));
```

Comment:
Looking at how `arithmeticElimination` works, I am not convinced that moving the substitutions in and then out is beneficial. Here it seems a reference parameter that serves as in-out parameter better captures the behaviour.

Reply:
Fixed.